byterun/intern.c

```diff
@@ -686,39 +763,54 @@ CAMLexport value caml_input_value_from_malloc(char * data, intnat ofs)
 /* [len] is a number of bytes */
 CAMLexport value caml_input_value_from_block(char * data, intnat len)
 {
-  uint32_t magic;
-  mlsize_t block_len;
+  struct marshal_header h;
   value obj;
 
   intern_input = (unsigned char *) data;
   intern_src = intern_input;
   intern_input_malloced = 0;
-  magic = read32u();
-  if (magic != Intext_magic_number)
-    caml_failwith("input_value_from_block: bad object");
-  block_len = read32u();
-  if (5*4 + block_len > len)
-    caml_failwith("input_value_from_block: bad block length");
-  obj = input_val_from_block();
+  caml_parse_header("input_value_from_block", &h);
+  if (h.header_len + h.data_len > len)
```

Comment:
Right, input_value and co are not protected against ill-formed inputs.  This is one example that could possibly go wrong, but I'm sure there are others.